File: sliding-window/longest-substring-without-repeating-characters.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        res,l = 0, 0
        count = set()
        for r in range(len(s)):
            if s[r] not in count:
                count.add(s[r])
                res = max(r-l+1, res)
            else:
                while l < r and s[r] in count:
                    count.remove(s[l])
                    l+=1
                count.add(s[r]) #be aware
        return res

File: sliding-window/longest-substring-without-repeating-characters.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        Set<Character> store = new HashSet<>();
        int maxLen = 0;
        int l =0;
        for (int r=0; r < s.length(); r++){
            while (store.contains(s.charAt(r))){
                store.remove(s.charAt(l));
                l+=1;
            }
            store.add(s.charAt(r));
            maxLen = Math.max(maxLen, r -l+1);

        }
        return maxLen;
        }
        
    }
